Update setup wizard - smart defaults with Jetpack

jeffstieler opened the pull request:
_Note: this still required quite a bit of styling, but is functional and can be reviewed._

This PR seeks to implement the designs in p7bbVw-28G-p2 and choose automatically configured payment gateways, shipping methods, and tax calculation as defaults in supported countries. It also promotes the benefits of Jetpack for store owners.

The old steps (Introduction, Pages, Location, Shipping, Payments, Theme, Ready) have been replace with:
* **Store setup** - New step. Store address, currency, product type, automatic page setup, tracking opt-in
* **Payment** - Reworked step with default to automatic Stripe account provisioning in supported countries
* **Shipping** - Reworked step to default to live rates when in supported countries, offers automatic zone creation and setup
* **Extras** - New step. Storefront installation and default to automated tax calculation in supported countries
* **Activate** - New step. Will install Jetpack and initiate connection process
* **Ready** - Reworked step. Cleaner design of the old step, removed tweets

jeffstieler committed:
Setup wizard: replaced “introduction” step with “store setup”.

New store setup step contains location, currency, page setup, and tracking opt-in. Form is display only.

jeffstieler committed:
Setup wizard: highlight first step in progress indicator.

jeffstieler committed:
Setup wizard: add “not right now” dashboard link on the initial wizard step.

jeffstieler committed:
Setup wizard: save settings from initial “store setup” step.

jeffstieler committed:
Setup wizard: remove (now) unused introduction step.

jeffstieler committed:
Setup wizard: replace “pages” step with “payments”.

jeffstieler committed:
Setup wizard: update Stripe gateway description, default it to enabled.

jeffstieler committed:
Setup wizard: check country for Stripe support before defaulting to it.

jeffstieler committed:
Setup wizard: differentiate between “in cart” and “manual” payment gateways.

This is in preparation to display the manual gateways behind a collapsable control.

jeffstieler committed:
Setup wizard: remove “skip” link from payments step.

jeffstieler committed:
Setup wizard: remove “location” step.

jeffstieler committed:
Setup wizard: add “activate” step, modeled after theme step.

Initial commit - needs styling love and Jetpack connection hookup.

jeffstieler committed:
Setup wizard: add “skip” link to activate step.

jeffstieler committed:
Setup wizard: refactor WooCommerce Services installer logic to be reusable.

jeffstieler committed:
Setup wizard: refactor plugin installation logic and use it in payment gateway save method.

jeffstieler committed:
Setup wizard: install WooCommerce Services if Stripe gateway is enabled.

jeffstieler committed:
Setup wizard: allow plugin install to happen immediately.

jeffstieler committed:
Setup wizard: install Jetpack and initiate connection flow when “saving” the activate step.

jeffstieler committed:
Setup wizard: show previously selected product type and remove shipping step when selling digital products only.

v18 committed:
Setup wizard: Add label for second address field

v18 committed:
Setup wizard: Move tracking checkbox to end of page

v18 committed:
Setup wizard: Fix Stripe description markup

v18 committed:
Setup wizard: Fix typo in PayPal Braintree description

v18 committed:
Setup wizard: Remove extra variable assignment

v18 committed:
Setup wizard: Add function to display service rows

- Update style
- Use new reusable class 'wc-wizard-services' instead of
'wc-wizard-payment-gateways'
- Update JS script for toggling checkboxes
- Update payment gateways: in-cart and manual
- Update POST variable names in save function

v18 committed:
Setup wizard: Update Stripe featured style

- filter in-cart payment gateways by featured
- display featured and others in separate lists

v18 committed:
Setup wizard: Add toggle for manual payment list

v18 committed:
Setup wizard: Fix typo

v18 committed:
Setup wizard: Fix list toggle name color

v18 committed:
Setup wizard: Add country-based intro to Shipping

v18 committed:
Setup wizard: Fix spacing

jeffstieler committed:
Setup wizard: fix PHP warning ‘undefined index’ in service item display method.

jeffstieler committed:
Setup wizard: basic layout changes for shipping step.

Still needs finer touches in styling and dynamic descriptions for shipping method options.

jeffstieler committed:
Setup wizard: update shipping step save method.

The shipping step now detects if live rates can be configured and allows the user to add shipping methods to a domestic and international zone. Note: flat rate shipping still needs an input for rate.

jeffstieler committed:
Setup wizard: replace “theme” step with “extras” step.

Presentation only. Determine need for a WooCommerce theme and support for automated taxes, showing options to enable each. If the store isn’t eligible for either, hide the step entirely.

jeffstieler committed:
Setup wizard: “extras” step save handler.

jeffstieler committed:
Setup wizard: remove now unused “theme” step.

jeffstieler committed:
Setup wizard: update “activate” step overall structure.

Change the Jetpack feature descriptions, graphic, placement of CTA button. Note: completely unstyled.

jeffstieler committed:
Setup wizard: update “ready” step overall structure.

Removed tweet, updated text. Note: completely unstyled.

jeffstieler committed:
Setup wizard: revert to combined country/state selection in initial step.

jeffstieler committed:
Setup wizard: fix UI block when HTML5 form validation fails.

jeffstieler committed:
Setup wizard: only toggle service enable control when its direct parent is clicked.

jeffstieler committed:
Setup wizard: add method-specific settings and descriptions to shipping step.

jeffstieler committed:
Setup wizard: enable tax calculation if automated taxes are enabled.

jeffstieler committed:
Setup wizard: simplify automated tax eligibility check.

Use the WC_Countries class to gather EU countries.

jeffstieler committed:
Setup wizard: only hook into method selection change event on shipping step.

jeffstieler committed:
Setup wizard: refactor shipping method form rendering to group inputs into an array.

jeffstieler committed:
Setup wizard: save shipping method settings chosen on shipping step.

jeffstieler committed:
Setup wizard: keep user in setup wizard in case of Jetpack connection error.

jeffstieler committed:
Setup wizard: fix form submit UI blocking when HTML5 form validation isn’t available.

jeffstieler committed:
Setup wizard: fix PHP warnings when saving shipping method settings.

jeffstieler committed:
Setup wizard: queue Jetpack installation ahead of payment gateways.

jeffstieler committed:
Setup wizard: set flag for queued Jetpack installs.

This sets up the ability to avoid installation race conditions.

jeffstieler committed:
Setup wizard: wait for in-progress Jetpack installation to complete before submitting form.

We need to know that Jetpack is active before moving to the next step. The previously logic fired off a synchronous installation that caused a race condition.

jeffstieler committed:
Setup wizard: set currency formatting options based on store location.

jeffstieler committed:
Setup wizard: remove now unused locale and tax interactions from client side script.

jeffstieler commented:
@mikejolley this is ready for a review 🔍 

@olaolusoga this is ready for styling ✨ 

cc: @v18

DanReyLop commented on assets/images/jetpack-green-logo.svg:
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Would be a good idea to run `svgo` on this: https://github.com/svg/svgo

DanReyLop commented on assets/js/admin/wc-setup.js:
```diff
@@ -1,99 +1,95 @@
+	}
+
+	$( '.button-next' ).on( 'click', function() {
+		var form = $( this ).parents( 'form' ).get( 0 );
```
You can use `$(this).closest('form');` instead: https://api.jquery.com/closest/

DanReyLop commented on assets/js/admin/wc-setup.js:
```diff
@@ -1,99 +1,95 @@
-		e.stopPropagation();
+	$( '.wc-wizard-services' ).on( 'change', '.wc-wizard-shipping-method-select .method', function( e ) {
+		var $zone = $( this ).parent( 'div' );
+		var selected_method = e.target.value;
```
What's the coding style? `snake_case` for variables, but `camelCase` for functions?

DanReyLop commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -48,42 +48,62 @@ public function admin_menus() {
-				'view'    => array( $this, 'wc_setup_pages' ),
-				'handler' => array( $this, 'wc_setup_pages_save' ),
+			'store_setup' => array(
+				'name'    => __( 'Store setup', 'woocommerce' ),
```
Should be `Store Setup` (title case)

DanReyLop commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -48,42 +48,62 @@ public function admin_menus() {
+			current_user_can( 'install_themes' ) &&
+			current_user_can( 'switch_themes' ) &&
+			! is_multisite() &&
+			! current_theme_supports( 'woocommerce' )
```
Why? Maybe the user has a theme that happens to somewhat support WooCommerce, but if he's building a store, shouldn't we recommend Storefront anyway?

jeffstieler replied:
This logic should be unchanged from the dedicated Storefront step.

DanReyLop commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -229,300 +262,132 @@ public function setup_wizard_content() {
 		$state          = WC()->countries->get_base_state();
 		$country        = WC()->countries->get_base_country();
 		$postcode       = WC()->countries->get_base_postcode();
+		$currency       = get_option( 'woocommerce_currency', 'GBP' );
```
Defaults to GBP?

jeffstieler replied:
This is the same logic as before.

DanReyLop commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -229,300 +262,132 @@ public function setup_wizard_content() {
+			<div style="width:33%; float:left;">
+				<label for="store_country_state"><?php esc_html_e( 'Country / State', 'woocommerce' ); ?></label>
+				<select id="store_country_state" name="store_country_state" style="width:100%;" required data-placeholder="<?php esc_attr_e( 'Choose a country&hellip;', 'woocommerce' ); ?>" aria-label="<?php esc_attr_e( 'Country', 'woocommerce' ) ?>" class="wc-enhanced-select">
+					<?php WC()->countries->country_dropdown_options( $country, $state ); ?>
```
There's a huge map returned by `WC_Countries->get_country_locale()` that tells you how things are named for each country (for example, in some countries `states` are called `provinces`, in USA the postcode is called `ZIP`, etc) and which parts are required/optional for a complete address.

Not super urgent, but we should use that data.

DanReyLop commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -229,300 +262,132 @@ public function setup_wizard_content() {
+				<input type="text" id="store_postcode" name="store_postcode" required value="<?php echo esc_attr( $postcode ); ?>" />
+			</div>
+			<div>
+				<label for="currency_code"><?php esc_html_e( 'Store currency', 'woocommerce' ); ?></label>
```
Would it make sense to smart default the currency based on the selected country?

jeffstieler replied:
I think this is a great idea but let's work on that separately. 👍

DanReyLop commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -578,55 +443,257 @@ protected function wc_setup_wcs_tout() {
+	 * @return array
+	 */
+	protected function get_wizard_shipping_methods( $country_code ) {
+		$shipping_methods = array(
```
`Local pickup` is missing, I don't know if it's intentional.

jeffstieler replied:
cc: @olaolusoga

olaolusoga replied:
It wasn't. thanks for the catch. We should include.

jeffstieler replied:
On second thought - we shouldn't include this. We're setting up either a zone that includes the entire country the store is located in, or all others. I don't think local pickup makes sense.

DanReyLop commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -578,55 +443,257 @@ protected function wc_setup_wcs_tout() {
+			<div class="shipping-method-description">
+			<?php foreach ( $shipping_methods as $method_id => $method ) : ?>
+				<p class="<?php echo esc_attr( $method_id ); ?>" <?php if ( $method_id !== $selected ) echo 'style="display:none"'; ?>>
+					<?php echo esc_html( $method['description'] ); ?>
```
Should this description be part of the `<option>` text? It feels like it would make things a bit clearer if you see all in the dropdown:
* Free Shipping - Don't charge for shipping
* etc

jeffstieler replied:
cc: @olaolusoga

olaolusoga replied:
Let's stick with the description on side. Though "Free shipping" can be described in 4 words, the same can't be said for live rates, flat rates, or local pickup.

DanReyLop commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -639,46 +706,85 @@ public function wc_setup_shipping() {
+		$intl_method      = sanitize_text_field( $_POST['shipping_zones']['intl']['method'] );
 		$weight_unit      = sanitize_text_field( $_POST['weight_unit'] );
 		$dimension_unit   = sanitize_text_field( $_POST['dimension_unit'] );
+		$existing_zones   = WC_Shipping_Zones::get_zones();
```
It's cleaner to just bail out of the function if `$existing_zones` isn't empty, instead of checking for it in every code block.

DanReyLop commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -578,55 +443,257 @@ protected function wc_setup_wcs_tout() {
+
+		$live_rate_carrier = $this->get_wcs_shipping_carrier( $country_code );
+
+		if ( false === $live_rate_carrier ) {
```
We should check here too if `current_user_can('install_plugins')`, for consistency, the same way you did for the payment methods.

jeffstieler committed:
Setup wizard: camelCase vs. snake_case in JavaScript.

jeffstieler committed:
Setup wizard: short circuit the shipping zone setup earlier.

jeffstieler committed:
Setup wizard: ensure that the current user can install plugins for any setup feature that requires it.

DanReyLop commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -747,74 +930,102 @@ protected function get_wizard_payment_gateways() {
+				<li class="wc-wizard-services-list-toggle">
+					<div class="wc-wizard-service-name">Manual Payments</div>
+					<div class="wc-wizard-service-description">
+						Collect payments from customers outside your online store.
```
Missing translations.

jeffstieler replied:
Fixed in https://github.com/woocommerce/woocommerce/pull/16755/commits/74d1b36f425e8a61fd8e09da5fa2fa358718f152.

DanReyLop commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -853,106 +1069,245 @@ public function wc_setup_payments_save() {
+		}
+
+		if ( $install_storefront ) {
+			wp_schedule_single_event( time() + 1, 'woocommerce_theme_background_installer', array( 'storefront' ) );
```
Using the "fake WordPress cron" for this seems hacky at best, and quite fragile at worst. It depends on a page being loaded after it's scheduled, and according to [this](https://wordpress.org/support/topic/scheduled-posts-still-not-working-in-282/#post-1175405) it's dependent on the server being properly configured.

I suggest doing this instead:
```
// flush any buffers and send the headers
while ( @ob_end_flush() );
flush();
// Prevent timeouts
ignore_user_abort(true);
// Silence any further print to stdout/stderr (I don't remember the function, ob_something())
// Install the plugin / theme syncronously
```

Same for the rest of the plugin / theme installs.

Commit:
Remove unused install_jetpack() function parameter.

Commit:
Removed unused "background" argument, since it's always true

Commit:
Added wrapper function for installing a theme

marcinbot committed:
Tailored the Jetpack step message to the services selected

marcinbot committed:
Fixed the way the Jetpack string is translated

mikejolley commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -229,300 +267,132 @@ public function setup_wizard_content() {
+					<option value="both" <?php selected( $product_type, 'both' ); ?>><?php esc_html_e( 'I plan to sell both', 'woocommerce' ); ?></option>
+				</select>
+			</div>
+			<p><?php esc_html_e( 'Transforming your site into an online store requires WooCommerce to create a few pages for you. These pages are: shop, cart, my account, and checkout.', 'woocommerce' ); ?></p>
```
Since this can no longer be skipped I don't think this is useful?

Wouldn't a checkbox to do this be sensible? Otherwise maybe dump it?

jeffstieler replied:
From a purely aesthetic standpoint I'd like to avoid adding a checkbox, though I'm curious - what's not useful about having store pages created?

mikejolley replied:
@jeffstieler The actual action of adding pages is useful. It's the message itself that is not IMO - it happens regardless. 

Previously you could skip the step.

v18 replied:
@mikejolley - I was thinking that the text may still be useful so that people aren't wondering where the pages came from after they're done the wizard? Could be surprising, no? cc @olaolusoga for a design perspective on this.

olaolusoga replied:
Agree @v18 . It's worth keeping for the cohort that require explicit informing about the installed pages. It's also there as a solution to the initial reason for the previous onboarding step (inform people of the installed pages). Though it can't be skipped, we could technically say the previous onboading was the same, since a WooCommerce store couldn't "exist" without the pages.

mikejolley replied:
@olaolusoga In previous version, you could skip that entire step and continue on with the rest of the wizard if the pages were not needed for any reason (rare?).

It's minor but worth pointing out this is now forced, thats why I thought this text was just an extra distraction if it's happening regardless.

olaolusoga replied:
Good point. Though it's forced, it does still feel like something we should inform ppl about, so they don't unintentionally delete the pages. 

This could be something we queue up for an A/B test. See if on-boarding is affected by keeping/removing it.

v18 replied:
@olaolusoga 👍 So, sounds like resolution is to leave off for now, and perhaps use it in an A/B test later.

mikejolley commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -578,55 +448,257 @@ protected function wc_setup_wcs_tout() {
+		$country_name   = WC()->countries->countries[ $country_code ];
+		$wcs_carrier    = $this->get_wcs_shipping_carrier( $country_code );
+
+		// TODO: determine how to handle existing shipping zones (or choose not to)
```
Was this missed?

jeffstieler replied:
Yes. Perhaps you can help make a decision on what the behavior should be here.

It's definitely a broken experience for any user re-running the setup wizard, but how common of an occurrence is that? Is this too much of an edge case to block the initial merge? _(e.g. can we do this later?)_

mikejolley replied:
@jeffstieler We do allow rerun but I think thats mainly for skipped. How about we disable the wizard by storing an option saying it was complete?

jeffstieler replied:
@mikejolley the wizard already removes the `install` notice on completion. Or are you specifically referring to shipping as "complete"?

mikejolley replied:
Actually talking about disabling it/preventing accessing it again via URL. We also show a button in the help tab to re-run. This should be removed if completed.

mikejolley commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -687,37 +803,81 @@ public function wc_setup_shipping_save() {
+	 * https://stripe.com/global
+	 */
+	protected function is_stripe_supported_country( $country_code ) {
+		$stripe_supported_countries = array(
```
Holding country names here seems pointless. They are not used.

jeffstieler replied:
Fixed in a97ecb712b5cbd1cb542984a69b1964006631c85.

mikejolley commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -687,37 +803,81 @@ public function wc_setup_shipping_save() {
+					'email' => array(
+						'label'       => __( 'Stripe email address', 'woocommerce' ),
+						'type'        => 'email',
+						'value'       => get_option( 'admin_email' ),
```
This should use the current user's user email first vs admin email. The admin may not be logged in or doing this.

jeffstieler replied:
Fixed in b8879dc520f9e19e3a85f2e4e70553918d38f589.

mikejolley commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -687,37 +803,81 @@ public function wc_setup_shipping_save() {
+			'stripe' => array(
+				'name'        => __( 'Stripe', 'woocommerce' ),
+				'image'       => WC()->plugin_url() . '/assets/images/stripe.png',
+				'description' => sprintf( __( '<p>Accept all major debit and credit cards from customers in 135+ countries on your site. <a href="%s" target="_blank">Learn more about Stripe</a>.</p><p class="payment-gateway-fee">Fee: 2.9%% + 30¢ per transaction</p>', 'woocommerce' ), 'https://wordpress.org/plugins/woocommerce-gateway-stripe/' ),
```
Localisation should not contain HTML.

jeffstieler replied:
I agree that this has too much HTML - are there examples in core of how to have inline links in translations?

jeffstieler replied:
Fixed in d8b7c21.

jeffstieler replied:
My motivation to keep the link in the HTML is proper RTL support. See this demo: https://codepen.io/anon/pen/JrKwvx

mikejolley commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -687,37 +803,81 @@ public function wc_setup_shipping_save() {
 			'paypal' => array(
 				'name'        => __( 'PayPal Standard', 'woocommerce' ),
 				'description' => __( 'Accept payments via PayPal using account balance or credit card.', 'woocommerce' ),
 				'image'       => '',
```
No image?

jeffstieler replied:
This is unchanged from the current wizard: https://github.com/woocommerce/woocommerce/blob/master/includes/admin/class-wc-admin-setup-wizard.php#L719

mikejolley commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -747,74 +930,102 @@ protected function get_wizard_payment_gateways() {
-			unset( $gateways['stripe'] );
-		}
+	public function is_featured_service( $service ) {
+		return isset( $service['featured'] ) && true === $service['featured'];
```
`! empty()` would be more efficient,

jeffstieler replied:
Fixed in 06cd161.

mikejolley commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -48,42 +48,69 @@ public function admin_menus() {
+
+	/**
+	 * The "automated tax" extra should only be shown if the current user can
+	 * install plugins and the store is in a supported coutnry
```
typo

jeffstieler replied:
Fixed in fa851fa.

jeffstieler committed:
Setup wizard: fix coding standard issues with “activate” step tailored messaging.

jeffstieler committed:
Setup wizard: remove contraction and add oxford comma to tailored “activate” step messaging.

jeffstieler committed:
Setup wizard: fix missing translations on payments step.

jeffstieler committed:
Setup wizard: clean up unused array keys when determining Stripe country support.

jeffstieler committed:
Setup wizard: use email address of the current user (not the admin) on payments and “ready!” steps.

jeffstieler commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -687,37 +803,81 @@ public function wc_setup_shipping_save() {
+						'placeholder' => __( 'Stripe email address', 'woocommerce' ),
+					),
+				),
+				'enabled' => $can_stripe,
```
I can switch it to opt-in. cc: @olaolusoga

jeffstieler replied:
Changed in 586c09793e2b6115cd48eee5d12e26217a7450a2.

jeffstieler committed:
Setup wizard: hide settings until a service is enabled.

jeffstieler committed:
Setup wizard: make Stripe an opt-in gateway, rename “manual payments” to “offline” payments to align with the header text.

jeffstieler committed:
Setup wizard: remove markup (where possible) from the translated Stripe service description.

jeffstieler commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -92,11 +119,19 @@ public function setup_wizard() {
 		}
 
+		// Whether or not there is a pending background install of Jetpack
+		$pending_jetpack = ! class_exists( 'Jetpack' ) && get_option( 'woocommerce_setup_queued_jetpack_install' );
```
They get stored as numeric string `0` and `1`.

jeffstieler committed:
Setup wizard: refactor featured service check to be more efficient.

Use `empty()` instead of `isset()` and a comparison.

jeffstieler committed:
Setup wizard: update comments to have ending punctuation and fix a few typos.

jeffstieler commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -92,11 +119,19 @@ public function setup_wizard() {
-		// Hide storefront step if using a WooCommerce theme or user cannot modify themes.
-		if ( ! current_user_can( 'install_themes' ) || ! current_user_can( 'switch_themes' ) || is_multisite() || current_theme_supports( 'woocommerce' ) ) {
-			unset( $default_steps['theme'] );
+		// Hide the extras step if this store/user isn't eligible for them
```
Fixed in fa851fa.

v18 committed:
Setup wizard: Add shipping list header

v18 committed:
Setup wizard: Fix extra padding in name column

v18 committed:
Setup wizard: Style shipping method description

v18 committed:
Setup wizard: Add shipping units descriptions

v18 committed:
Setup wizard: Style shipping unit selection

jeffstieler committed:
Setup wizard: update the copy in several places.

v18 committed:
Setup wizard: Fix thick border in toggle list

v18 committed:
Setup wizard: Show correct shipping text on select

v18 committed:
Setup wizard: Style Jetpack feature sections

jeffstieler committed:
Setup wizard: change copy to “recommended” extras.

v18 committed:
Setup wizard: Style Jetpack logo

v18 committed:
Setup wizard: Update button style

v18 committed:
Setup wizard: Update page padding

v18 committed:
Setup wizard: Style store info form

v18 committed:
Setup wizard: Style email input for payment service

v18 committed:
Wizard: Remove inline CSS for unit dropdowns

jeffstieler commented:
We're getting close on this @mikejolley - styling is being worked here: https://github.com/woocommerce/woocommerce/tree/add/new-wizard-css

v18 committed:
Wizard: Style shipping method selection

v18 committed:
Wizard: Add placeholder to flat rate cost input

Commit:
Added logic to install the plugins and themes asynchronously but without using wp-cron

v18 committed:
Wizard: Make payment service email required

v18 committed:
Wizard: style newsletter form

jeffstieler committed:
Setup wizard: use form-specific selector to trigger Jetpack connect behavior.

v18 committed:
Wizard: Style next actions section

v18 committed:
Wizard: Style footnote on last page

v18 commented on assets/js/admin/wc-setup.js:
```diff
@@ -1,99 +1,97 @@
+			return true;
+		}
+
+		e.preventDefault();
```
Should we prevent the default in the beginning of the function?

jeffstieler committed:
Setup wizard: add TOS blurb to the “activate” step.

mikejolley commented:
These are the final styling things that need attention.

cc @olaolusoga @v18 @jeffstieler 

Additional note: Not sure there is time now, but an editorial review would be great.

1. **Store setup.** 

Last p:last-child needs no margin bottom, and that final section with the currency could be better split from the above address. Those bottom fields could even be side by side?

<img width="801" alt="2017-09-27 at 11 22" src="https://user-images.githubusercontent.com/90977/30908675-3954d92c-a376-11e7-92a6-c8a143433e0d.png">

The biggest problem here is step 1 is a wall of fields.  I think a little spacing may also make this look less daunting. Or a graphic. Just something that makes it engaging :p 

The heading could be removed and text tweaked to remove overall noise:

<img width="768" alt="2017-09-27 at 11 36" src="https://user-images.githubusercontent.com/90977/30909176-223d2bf2-a378-11e7-972e-dd70705c6337.png">

Also, the "I plan to sell both" option is not great. To know what the options are you need to click the dropdown - there is no indication this option is talking about digital vs physical goods. At the very least, the default value should be "I plan on selling both digital and physical products".

2. **Payments**

Position is off:

<img width="460" alt="2017-09-27 at 11 38" src="https://user-images.githubusercontent.com/90977/30909284-724d045a-a378-11e7-938d-cebf66987b41.png">

The logos are also not top aligned like everything else.

3. **Shipping**

> You can choose which countries you'll be shipping to and with which methods. We've started you up with shipping to United Kingdom (UK) and the rest of the world.

I'd suggest:

> You can choose which countries you'll be shipping to and with which methods. To get started, we've set you up with shipping inside and outside of the United Kingdom (UK).

Note; core has a method for the 'THE' prefix in the countries class, since some county names need a prefix for good grammar.

4. **Extras**

> Make the most of your store.

Is this description needed? Adds no value.

5. **Activate**

Not really happy with this step.

1. Option detection is broken. If you go back a step and uncheck taxes, the option will not be updated to false.
2. **Services and Jetpack** is not a requirement for Stripe. This wording makes it sound otherwise. Easier setup for Stripe is a bonus but I don't feel it needs a specific mention.
3. There are 7 jetpack strings to localise based on selections. Could cut down to one string with replacement: `Your store is almost ready! To activate services (X) just connect with Jetpack.`
4. If the user **has not opted** into services, this step shouldn't really be shown, but if it's being pushed anyway, **make the skip this step a secondary action rather than give the appearance that this is required.** 

<img width="672" alt="2017-09-27 at 12 01" src="https://user-images.githubusercontent.com/90977/30910193-aeb5863a-a37b-11e7-9f5b-bd4b86b65e5d.png">

If services were opted into, this isn't a problem so make it conditional.

FWIW this is what it looks like without opt-in to anything:

<img width="765" alt="2017-09-27 at 12 07" src="https://user-images.githubusercontent.com/90977/30910379-7a924892-a37c-11e7-95bb-e56e4957c2d7.png">

So this needs IMO:

1. Prominent skip link
2. "Bonus" can be removed from string.
3. @jeffstieler Can TOS be moved to bottom?
4. "Connect your store to Jetpack" -> "Connect your store to Jetpack for added functionality "

6. Final step.

Remove text "Your WooCommerce store is ready to go.". It's the same as the heading pretty much.

<img width="433" alt="2017-09-27 at 11 52" src="https://user-images.githubusercontent.com/90977/30909829-62b55b08-a37a-11e7-8c84-250ec453345a.png">

> Return to the WP Admin

Change to:

> Return to your dashboard

v18 committed:
Wizard: Fix chevron placement in toggle list

v18 committed:
Wizard: Fix logo placement in services list

v18 committed:
Wizard: Update shipping info text

v18 committed:
Wizard: Remove superfluous Extras description

v18 committed:
Wizard: Remove superfluous description in last step

v18 committed:
Wizard: Update 'WP Admin' to 'dashboard'

v18 commented:
Thank you @mikejolley!

> Additional note: Not sure there is time now, but an editorial review would be great.

We've been incorporating editorial comments as we went along, fwiw - working from a shared google doc. Let us know if you see any more copy issues.

> 1. Store setup.
> Last p:last-child needs no margin bottom, and that final section with the currency could be better split from the above address. Those bottom fields could even be side by side?

I was following @olaolusoga's designs there, where there's extra space between the two sections. Here's the original design:

![location-form](https://user-images.githubusercontent.com/11487924/30916677-a7043932-a3a2-11e7-8977-dfb2abb57b54.jpg)

@olaolusoga what do you think about the current design as it stands? Here's a full image of it:

<img width="763" alt="first-page-sept-27" src="https://user-images.githubusercontent.com/11487924/30916750-d1e320be-a3a2-11e7-8542-de9324387b47.png">

> 2. Payments

Fixed the chevron in 263a471c40724eb74de0ab0bc126a8065ac0e830

Also fixed the logos on that page in 562f37520df6c53eb06b8313a3ee041dcebfd0e7

> 3. Shipping

Updated text + added 'the' prefix in a09b7f167019e231c978c319960281767ae8f616

> 4. Extras

Removed in d6c94e4837770923fbe2610747dc47d111082345. cc @olaolusoga for an opinion on the new page. Here's an example of what it might look like:

<img width="761" alt="screen shot 2017-09-27 at 5 27 04 pm" src="https://user-images.githubusercontent.com/11487924/30919088-1a7ecbd8-a3a9-11e7-8333-80018d52b1de.png">

> 6. Final step.

Removed "Your WooCommerce store is ready to go" in bd98a7098d0cb57972317cca3ba521836e4a5946 and updated wp admin wording in 0b2bad81229a2e857cf0e45910b73f541af1ebfd

Looking at the JP issues now.

mikejolley commented:
@v18 

> I was following @olaolusoga's designs there, where there's extra space between the two sections. Here's the original design:

The margin bottom part is an easy fix even if the placement is not changed. Note the margin bottom on the p wrapping the button, compared to the comps.

v18 committed:
Wizard: Consolidate button styles in one rule

v18 committed:
Wizard: Avoid too much space after last paragraph

v18 commented:
@mikejolley - ah, I see what you mean now. Fixed the bottom margin for the paragraph in d45ae3877091514f69809b09e7efdb78a39fed95. Waiting for @olaolusoga  for form placement ideas.

jeffstieler commented:
> **Activate step**
> 
> Option detection is broken. If you go back a step and uncheck taxes, the option will not be updated to false.

This sounds like a bug in the `_save()` handlers, not specific to the activate step. @v18 is working this now.

> Services and Jetpack is not a requirement for Stripe. This wording makes it sound otherwise. Easier setup for Stripe is a bonus but I don't feel it needs a specific mention.

Because of the deferred account setup (which does require Services) we're triggering the Jetpack+Services install based on selecting Stripe. The new email field is used for the account setup. Since we're doing that, it seems worth a mention. Is there a middle ground for this?

> There are 7 jetpack strings to localise based on selections. Could cut down to one string with replacement: Your store is almost ready! To activate services (X) just connect with Jetpack.

👍 

> @jeffstieler Can TOS be moved to bottom?

Actually, it can't. According to @jennyzhu it has to appear **above** the button to connect.

> "Connect your store to Jetpack" -> "Connect your store to Jetpack for added functionality "

👍

mikejolley commented:
@jeffstieler I'd tweak the styles then. It's also below the button ;)

> Because of the deferred account setup (which does require Services) we're triggering the Jetpack+Services install based on selecting Stripe. The new email field is used for the account setup. Since we're doing that, it seems worth a mention. Is there a middle ground for this?

I feel like this should be used only as a bonus if they opt-in to services, or opt-in to Jetpack. If they do - great! Let's help them! Otherwise, it's extra steps for seemingly little added benefit? Stripe standalone does not need 2 extra plugins to function. Thats why I feel this step is misleading.

mikejolley commented:
@v18 Remove the tracking option from your DB - tracking is not styled https://dl.dropboxusercontent.com/s/qvubfggck3mhm0c/2017-09-27%20at%2017.48.png

Option name is `woocommerce_allow_tracking`

mikejolley commented:
@v18 (sorry) also that p style change I think needs to be last-child, not last-of-type, otherwise it affects margins further up:

<img width="726" alt="2017-09-27 at 18 00" src="https://user-images.githubusercontent.com/90977/30926665-c45001dc-a3ad-11e7-8391-abcecabd2d27.png">

v18 committed:
Wizard: Correctly save taxes opt-in status

- Make sure to save the option every time, even if not selected
- Show the corresponding state in the toggle

v18 commented:
> @v18 (sorry) also that p style change I think needs to be last-child, not last-of-type, otherwise it affects margins further up:

Nice catch. We needed last-of-type because sometimes there were elements after the last paragraph. Will fix 👍

v18 commented:
Some Github Desktop app magic just happened :) Re-opened this PR to fix.

jeffstieler committed:
Setup wizard: consolidate translations for Jetpack enabled features list as much as possible.

jeffstieler commented:
@claudiulodro consolidated the translation strings as much as I see possible in https://github.com/woocommerce/woocommerce/pull/16755/commits/7b9b09e9ad073ae72c0d5cb30a0a8ffcc137b76d - figuring that building a comma-separated feature list programmatically breaks RTL translation.

jeffstieler committed:
Setup wizard: default Stripe gateway on, swap email input for opt-in account creation.

v18 committed:
Wizard: Save payments enabled status

- toggle payment option on if its setting was previously saved,
  or if it's listed as to be toggled on be default
- check if setting is saved as enabled in options

jeffstieler committed:
Setup wizard: for non-Services merchants, mention that Jetpack is for “extra” features on the activate step.

v18 commented:
> 5. Activate step
> Option detection is broken. If you go back a step and uncheck taxes, the option will not be updated to false.

I fixed it for payments & taxes in b772b002a9c908844536e93b98faef0b5ddcd6e5 and 8c2a2860e3e30d5baed09f4540d1420a3d9274e6

v18 committed:
Wizard: Move Jetpack TOS above button

v18 committed:
Wizard: Fix bad email check for Stripe status

We are no longer asking for the user's email or
saving it, so it shouldn't be checked against.

jeffstieler committed:
Setup wizard: use a setup wizard specific ID for the Jetpack connection URL.

claudiulodro committed:
Bring up-to-date with master and recompile css

claudiulodro commented:
Merging this in with the understanding that there will be some tweaks and improvements during RC period.

v18 commented:
@claudiulodro for future reference - would you like the compiled CSS changes to come at the same time as the related .scss or in one big commit at the end? Or something else?

claudiulodro commented:
@v18 doesn't matter how it's done as long as the css is compiled before the PR is merged. Whichever works best for you. The important thing is that the master branch should always have the latest compiled CSS.

v18 commented:
Got it, thanks! Was planning on adding a commit with all the compiled CSS at once - sounds like it was a good plan :)
